security/pkg/nodeagent/cache/monitoring.go

```diff
@@ -55,3 +59,17 @@ func init() {
 		numFileSecretFailures,
 	)
 }
+
+func recordCertExpiry(resourceName string, expiry time.Time) {
+	labels := map[string]string{
+		"resource_name": resourceName,
+	}
+	monitoring.NewDerivedGaugeWithLabels(
```

Comment:
what happens if this is called more than once with different resourceNames?

Reply:
the metric will be updated.
```golang
	recordCertExpiry(item.ResourceName, item.ExpireTime)
	recordCertExpiry(item.ResourceName+"bak", item.ExpireTime)
```
output:
```console
# HELP istio_agent_cert_expiry_seconds The left time, in seconds, when cert chain will expire. A negative value indicates the cert is expired.
# TYPE istio_agent_cert_expiry_seconds gauge
istio_agent_cert_expiry_seconds{resource_name="defaultbak"} 86355.853291707

```

Reply:
Yes. I guess I'm asking if there is ever a valid case for two cert expiry metrics in the same process.

Reply:
IIRC, only `default` cert, @howardjohn can you confirm this?

Reply:
We support file mounted ones, so may be unbounded number actually. Also ROOTCA may be something we want to measure as well..?

Reply:
I think the workaround we want is something like: https://github.com/istio/pkg/pull/483/files#r810176904

Reply:
https://github.com/census-instrumentation/opencensus-go/blob/49838f207d61097fc0ebb8aeef306913388376ca/metric/registry.go#L237

the `registry` use `metric.Name` as key, I think we can not have multi `DerivedGauge` (same name but different in labels) before we modify the implement of upstream.

Reply:
What is current status? Last one wins?

If so that is not a great option IMO. Either we only do `default`, do `default`+`ROOTCA`, only do the oldest, or do all of them. But doing the last one will give unexpected results

Reply:
@zirain, I was OOO so I may have lost context, but the metric.Name as the key shouldn't be an issue. If the function is applied with labels, it should be possible to have multiple values. I'll try to add the functionality with tests and ping back here.

Reply:
> What is current status? Last one wins?
> 
> If so that is not a great option IMO. Either we only do `default`, do `default`+`ROOTCA`, only do the oldest, or do all of them. But doing the last one will give unexpected results

the last win, but right now it will only be called at `registerSecret`.

Reply:
> @zirain, I was OOO so I may have lost context, but the metric.Name as the key shouldn't be an issue. If the function is applied with labels, it should be possible to have multiple values. I'll try to add the functionality with tests and ping back here.

if we can make `DerivedGauge` support metrics with same name and labels keys but different label values, everything will be more easier.

Reply:
@zirain @howardjohn please take a look at https://github.com/istio/pkg/pull/490. i think that (or similar) should allow us to define the metric once and have multiple calls like:

```golang
certExpirySeconds = monitoring.NewDerivedGauge(
       "cert_expiry_seconds",
       "The left time, in seconds, when cert chain will expire. ",
       monitoring.WithLabelKeys("resource_name"),
)
...
certExpirySeconds.ValueFrom(func() float64 { return time.Until(expiry).Seconds() }, resourceName)
```